src/why3_tactic.ml

```diff
@@ -0,0 +1,198 @@
+
+(** [create tbl hyps term] adds all the constants of [tbl] in a new task, also
+    declare the hypotheses of [hyps] as axioms and register the term [term] as
+    a Why3 goal. *)
+let create : cnst_table -> Why3.Term.term StrMap.t -> Why3.Term.term ->
+               Why3.Task.task = fun tbl hyps term ->
+  (* Add the declaration of every constant. *)
+  let fn tsk (_,t) = Why3.Task.add_param_decl tsk t in
+  let tsk = List.fold_left fn None tbl in
+  (* Add the declaration of every hypothesis. *)
+  let fn name t tsk =
+    let axiom = Why3.Decl.create_prsymbol (Why3.Ident.id_fresh name) in
+    Why3.Task.add_prop_decl tsk Why3.Decl.Paxiom axiom t
+  in
+  let tsk = StrMap.fold fn hyps tsk in
+  (* Add the goal itself. *)
+  let goal = Why3.Decl.create_prsymbol (Why3.Ident.id_fresh "main_goal") in
+  Why3.Task.add_prop_decl tsk Why3.Decl.Pgoal goal term
+
+(** [prover pos prv] search and return the prover called [prv]. *)
+let prover : Pos.popt -> string -> Why3.Whyconf.config_prover = fun pos prv ->
+  (* Filter the set of why3 provers. *)
+  let fp = Why3.Whyconf.parse_filter_prover prv in
+  (* Get the set of provers. *)
+  let provers = Why3.Whyconf.filter_provers why3_config fp in
+  (* Display a message if we did not find a matching prover. *)
+  if Why3.Whyconf.Mprover.is_empty provers then
+    Console.fatal pos "[%s] not installed or not configured" prv;
+  (* Return the prover configuration. *)
+  snd (Why3.Whyconf.Mprover.max_binding provers)
+
+(** [handle prover_name ss tac ps g] FIXME *)
+let handle prover_name ss tac ps g =
+  (* Get the goal to prove. *)
+  let (hypotheses, trm) = Proof.Goal.get_type g in
+  (* Get the default or the indicated name of the prover. *)
+  let prover_name = Option.get prover_name Timed.(!default_prover) in
+  (* Translate from lambdapi to why3 terms. *)
+  let (constants, hyps, why3term) =
+    translate tac.Pos.pos ps.Proof.proof_builtins (hypotheses, trm)
+  in
+  (* Create a new task that contains symbols, axioms and the goal. *)
+  let tsk = create constants hyps why3term in
+  (* Call the prover named [prover_name] and get the result. *)
+  let prover_result =
+    let prv = prover tac.pos prover_name in
+    let limit_time = !prover_timeout in
+    let limit = {Why3.Call_provers.empty_limit with limit_time} in
+    let command = prv.Why3.Whyconf.command in
+    (* Load the config prover [prv] in current environment, return the driver
+       of the prover. *)
+    let driver =
+      try Why3.Whyconf.(load_driver why3_main why3_env prv.driver [])
+      with e ->
+        Console.fatal tac.pos "Failed to load driver for %s: %a"
+          prv.prover.prover_name Why3.Exn_printer.exn_printer e
+    in
+    Why3.Call_provers.wait_on_call
+      (Why3.Driver.prove_task ~limit ~command driver tsk)
+  in
+  (* Check that the prover succeeded to prove the goal. *)
+  if not (Why3.Call_provers.Valid = prover_result.pr_answer) then
+    Console.fatal tac.pos "%s did not found a proof@." prover_name;
```

Comment:
For `run_task`, by + `prover` you mean I should inline `prover` in this new function ?

Reply:
Yes, exactly.

Reply:
The reason why I think it should be inlined is because the "middle interface" (the triple returned by `translate`) is not really interesting to us, so it is not worth having to document it.

Reply:
Check dad84a5 if it is better like this.